### PR TITLE
Change output dataloader

### PR DESCRIPTION
**Pull request by evvahu**
In this PR, the return variables are changed in the DataLoader so that one training script can be used for all kinds of classifiers.
The following parts of the code had to be changed:
- data loader: return list changed to return dictionary
- all classifiers (the forward method)
- all tests connected to the parts metnioned above
- train.py

train_rnn.py has not been removed. Also, the phrase_context classifier does not work yet because of the additional parameter  (_device_) in the forward method.

### CODE REVIEW

File: scripts/basic_twoword_classifier.py [hunk 1]
```diff
@@ -17,15 +17,18 @@ def __init__(self, input_dim, hidden_dim, label_nr, dropout_rate):
...
         this function takes two words, concatenates them and applies a non-linear matrix transformation (hidden layer)
         Its output is then fed to an output layer. Then it returns the concatenated and transformed vectors.
         :param word1: the first word of size batch_size x embedding size
         :param word2: the first word of size batch_size x embedding size
         :return: the transformed vectors after output layer
         """
-        word_composed = comp_functions.concat(word1, word2, axis=1)
+
+        word_composed = comp_functions.concat(batch["w1"], batch["w2"], axis=1)
+        #word_composed = comp_functions.concat(word1, word2, axis=1)
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py:31]: #word_composed = comp_functions.concat(word1, word2, axis=1)
remove old code

File: scripts/data_loader.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
 import somajo
 import numpy as np
 from scripts import BertExtractor, StaticEmbeddingExtractor
+from types import SimpleNamespace
```
Comment (Blubberli) [inline scripts/data_loader.py:10]: from types import SimpleNamespace
Remove unused import

File: scripts/data_loader.py [hunk 3]
```diff
@@ -145,7 +146,13 @@ def lookup_embedding(self, words):
...
         word1_embeddings = self.lookup_embedding(self.word1)
         word2_embeddings = self.lookup_embedding(self.word2)
-        return [[word1_embeddings[i], word2_embeddings[i], self.labels[i]] for i in range(len(self.labels))]
+
+        return [{"w1": word1_embeddings[i], "w2": word2_embeddings[i], "l": self.labels[i]} for i in range(len(self.labels))]
+
+        #return [SimpleNamespace(word1_embs=word1_embeddings[i],
+         #                       word2_embs=word2_embeddings[i],
+          #                      labs=self.labels[i]) for i in range(len(self.labels))]
+        #return [[word1_embeddings[i], word2_embeddings[i], self.labels[i]] for i in range(len(self.labels))]
```
Comment (Blubberli) [inline scripts/data_loader.py:156]: remove old / unused code

File: scripts/data_loader.py [hunk 4]
```diff
@@ -199,9 +206,19 @@ def populate_samples(self):
...
+                                "seq":sequences[i],
+                                "seq_lengths": sequence_lengths[i],
+                                "l": self.labels[i]} for i in range(len(self.labels))]
+        #return [SimpleNamespace(word1_embs = word1_embeddings[i],
+         #                   word2_embs= word2_embeddings[i],
+          #                  seqs = sequences[i],
+           #                 labs = self.labels[i]) for i in range(len(self.labels))]
+
+        #return [[word1_embeddings[i], word2_embeddings[i], sequences[i], sequence_lengths[i], self.labels[i]] for i in
+                #range(len(self.labels))]
```
Comment (Blubberli) [inline scripts/data_loader.py:222]: remove old / unused code

File: scripts/data_loader.py [hunk 5]
```diff
@@ -276,8 +293,15 @@ def populate_samples(self):
...
-                range(len(self.labels))]
+        return [{"w1": word1_embeddings[i],
+                                "w2" : word2_embeddings[i],
+                                "seq":sequences[i],
+                                "seq_lengths": sequence_lengths[i],
+                                "l": self.labels[i]} for i in range(len(self.labels))]
+
+
+        #return [[word1_embeddings[i], word2_embeddings[i], sequences[i], sequence_lengths[i], self.labels[i]] for i in
+         #       range(len(self.labels))]
```
Comment (Blubberli) [inline scripts/data_loader.py:305]: siehe oben ^^

File: scripts/data_loader.py [hunk 6]
```diff
@@ -199,9 +206,19 @@ def populate_samples(self):
         sequences = nn.utils.rnn.pad_sequence(batch_first=True, sequences=sequences, padding_value=0.0)
 
         sequence_lengths = np.array([len(words) for words in self.sentences])
-
-        return [[word1_embeddings[i], word2_embeddings[i], sequences[i], sequence_lengths[i], self.labels[i]] for i in
-                range(len(self.labels))]
+        list_of_namespaces = []
```
Comment (Blubberli) [inline scripts/data_loader.py:209]: list_of_namespaces = []
This is never used

File: tests/test_basic_twoword_classifier.py [hunk 7]
```diff
@@ -12,19 +14,26 @@ class BasicTwoWordClassifierTest(unittest.TestCase):
     """
 
     def setUp(self):
-        self.word1 = torch.from_numpy(np.array([[1, 0, 0]], dtype='float32'))
-        self.word2 = torch.from_numpy(np.array([[0, 1, 0]], dtype='float32'))
-        self.input_dim = 6
+        #self.word1 = torch.from_numpy(np.array([[1, 0, 0]], dtype='float32'))
+        #self.word2 = torch.from_numpy(np.array([[0, 1, 0]], dtype='float32'))
+        self._data_path = pathlib.Path(__file__).parent.absolute().joinpath("data_multiclassification/test.txt")
```
Comment (Blubberli) [inline tests/test_basic_twoword_classifier.py:19]: siehe oben

File: scripts/basic_twoword_classifier.py [hunk 8]
```diff
@@ -17,15 +17,18 @@ def __init__(self, input_dim, hidden_dim, label_nr, dropout_rate):
         self._hidden_layer = nn.Linear(input_dim, hidden_dim)
         self._output_layer = nn.Linear(hidden_dim, label_nr)
         self._dropout_rate = dropout_rate
-    def forward(self, word1, word2, training=True):
+    #def forward(self, word1, word2, training=True):
+    def forward(self, batch, training=True):
```
Comment (Blubberli) [inline scripts/basic_twoword_classifier.py:21]: #def forward(self, word1, word2, training=True):
remove

File: scripts/phrase_context_classifier.py [hunk 9]
```diff
@@ -29,12 +29,12 @@ def __init__(self, embedding_dim, hidden_size, num_layers, forward_hidden_dim, l
         self._hidden_layer = nn.Linear(2 * embedding_dim + hidden_size * 2, forward_hidden_dim)
         self._output_layer = nn.Linear(forward_hidden_dim, label_nr)
 
-    def forward(self, word1, word2, context, context_lengths, training, device):
+    def forward(self, batch, training, device):
         # context size = batchsize x max len x embedding dim
         # convert the padded context into a packed sequence such that the padded vectors are not shown to the LSTM
         # context_packed = sum of all seq lenghts, embedding_dim
         # batch_sizes = column-wise (how many real elements do I have?)
-        context_packed = nn.utils.rnn.pack_padded_sequence(context, context_lengths, batch_first=True,
```
Comment (Blubberli) [inline scripts/phrase_context_classifier.py:37]: maybe we can add an assert statement here to ensure that "seq" and "seq_lengths" are in the dictionary ..
assert "seq" in batch.keys() and "seq_lengths" in batch.keys(), "no sequence and sequence length in bacht" o.ä dann weiß man gleich dass man das falsche Datenset geladen hat

**Review by Blubberli — CHANGES_REQUESTED** [review]
Nice, works well. Be brave and remove old code after you are done instead of leaving it commented out ;) . Can you update the test_phrase_and_context_dataset.p? I think that one is missing...
